linera-views/src/value_splitting.rs

```diff
@@ -0,0 +1,378 @@
+        let mut big_key_segment = key.to_vec();
+        let mut bytes = bcs::to_bytes(&index)?;
+        bytes.reverse();
+        big_key_segment.extend(bytes);
+        Ok(big_key_segment)
+    }
+
+    fn read_index_from_key(key: &[u8]) -> Result<u32, K::Error> {
+        let len = key.len();
+        if len < 4 {
+            return Err(DatabaseConsistencyError::TooShortKey.into());
+        }
+        let mut bytes = key[len - 4..len].to_vec();
+        bytes.reverse();
+        Ok(bcs::from_bytes::<u32>(&bytes)?)
+    }
+
+    fn get_initial_count_first_chunk(count: u32, first_chunk: &[u8]) -> Result<Vec<u8>, K::Error> {
+        let mut bytes = bcs::to_bytes(&count)?;
+        bytes.reverse();
+        let mut value_ext = Vec::new();
+        value_ext.extend(bytes);
+        value_ext.extend(first_chunk);
+        Ok(value_ext)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use linera_views::{
+        batch::Batch,
+        common::KeyValueStoreClient,
+        memory::{create_test_client_internal, MemoryClientInternal},
+        value_splitting::KeyValueStoreClientBigValue,
+    };
+    use rand::{Rng, SeedableRng};
+
+    // The key splitting means that when a key is overwritten
+    // some previous segments may still be present.
+    #[tokio::test]
+    #[allow(clippy::assertions_on_constants)]
+    async fn test_value_splitting1_testing_leftovers() {
+        let client = create_test_client_internal();
+        const MAX_LEN: usize = MemoryClientInternal::MAX_VALUE_SIZE;
+        assert!(MAX_LEN > 10);
+        let big_client = KeyValueStoreClientBigValue::new(client.clone());
+        let key = vec![0, 0];
+        // Write a key with a long value
+        let mut batch = Batch::new();
+        let value = Vec::from([0; MAX_LEN + 1]);
+        batch.put_key_value_bytes(key.clone(), value.clone());
+        big_client.write_batch(batch, &[]).await.unwrap();
+        let value_read = big_client.read_key_bytes(&key).await.unwrap();
+        assert_eq!(value_read, Some(value));
+        // Write a key with a smaller value
+        let mut batch = Batch::new();
+        let value = Vec::from([0, 1]);
+        batch.put_key_value_bytes(key.clone(), value.clone());
+        big_client.write_batch(batch, &[]).await.unwrap();
+        let value_read = big_client.read_key_bytes(&key).await.unwrap();
+        assert_eq!(value_read, Some(value));
+        // Two segments are present even though only one is used
+        let keys = client.find_keys_by_prefix(&[0]).await.unwrap();
+        assert_eq!(keys, vec![vec![0, 0, 0, 0, 0], vec![0, 0, 0, 0, 1]]);
+    }
+
+    #[tokio::test]
+    async fn test_value_splitting2_testing_splitting() {
+        let client = create_test_client_internal();
+        const MAX_LEN: usize = MemoryClientInternal::MAX_VALUE_SIZE;
+        let big_client = KeyValueStoreClientBigValue::new(client.clone());
+        let key = vec![0, 0];
+        //
```

Comment:
Did you mean to add a comment here? If it's just a separator I'd make it a blank line.

Reply:
I added a comment.